include/wil/cppwinrt_helpers.h

```diff
@@ -221,11 +221,27 @@ namespace wil
     /// @cond
     namespace details
     {
-        template<typename T> struct is_vector : std::bool_constant<false> {};
-        template<typename Q> struct is_vector<winrt::Windows::Foundation::Collections::IVector<Q>> : std::bool_constant<true> {};
-        template<typename Q> struct is_vector<winrt::Windows::Foundation::Collections::IVectorView<Q>> : std::bool_constant<true> {};
-        template<typename T> struct is_iterator : std::bool_constant<false> {};
-        template<typename Q> struct is_iterator<winrt::Windows::Foundation::Collections::IIterator<Q>> : std::bool_constant<true> {};
+        template<typename T> struct is_winrt_vector_like {
+        private:
+            template <typename U,
+                typename = decltype(std::declval<U>().GetMany(std::declval<U>().Size(),
+                    winrt::array_view<decltype(std::declval<U>().GetAt(0))>{}))>
+            static constexpr bool get_value(int) { return true; }
+            template <typename> static constexpr bool get_value(...) { return false; }  
+        public:
+            static constexpr bool value = get_value<T>(0);                                                                                                                         
+        };

```

Comment:
For future reference, here's a version that supports customizing the type of the vector and the allocator. We probably don't need it, but just saving here for posterity.

```cpp
        template<typename T, typename ElementType>
        class is_winrt_vector_like
        {
            template <typename U, typename = decltype(std::declval<U>().GetMany(0, winrt::array_view<ElementType>{})) >
            static constexpr bool get_value(int) { return true; }
            template <typename>
            static constexpr bool get_value(...) { return false; }

        public:
            static constexpr bool value = get_value<T>(0);
        };
    }

    template<typename ElementType = void, typename Allocator = void, typename T,
        typename ActualElementType = std::conditional_t<std::is_same_v<ElementType, void>, std::decay_t<decltype(std::declval<T>().GetAt(0))>, ElementType>,
        typename ActualAllocator = std::conditional_t<std::is_same_v<Allocator, void>, std::allocator<ActualElementType>, Allocator>,
        typename std::enable_if_t<details::is_winrt_vector_like<T, ActualElementType>::value, int> = 0>
        auto to_std_vector(T const& v, ActualAllocator const& alloc = ActualAllocator())
    {
        std::vector<ActualElementType, ActualAllocator> result(alloc);
        auto size = v.Size();
        if (size > 0)
        {
            result.resize(size + 1, winrt_empty_value<ActualElementType>());
            auto actual = v.GetMany(0, result);
            if (actual > size)
            {
                throw winrt::hresult_changed_state();
            }
            result.erase(result.begin() + actual, result.end());
        }
        return result;
    }
```

The weirdness with the template types is to allow you to write `to_std_vector<MyType, MyAllocator>(winrtVector)`.